msal/application.py

```diff
@@ -862,16 +863,16 @@ def _acquire_token_by_username_password_federated(
         GRANT_TYPE_SAML1_1 = 'urn:ietf:params:oauth:grant-type:saml1_1-bearer'
         grant_type = {
             SAML_TOKEN_TYPE_V1: GRANT_TYPE_SAML1_1,
-            SAML_TOKEN_TYPE_V2: self.client.GRANT_TYPE_SAML2,
+            SAML_TOKEN_TYPE_V2: self._get_client().GRANT_TYPE_SAML2,
```

Comment:
For this one and the next one and the `encode_saml_assertion`, we can probably go with a different calling pattern like this:
```suggestion
            SAML_TOKEN_TYPE_V2: Client.GRANT_TYPE_SAML2,
```

Actually, since you missed line 874's `self.client.encode_saml_assertion`, it seems your testing strategy is to "run it (initially by test automation, and now by manual test), and see if it is broken, if so, fix it". That is not enough. We should aim for have a white-box decision BEFORE attempting to run the test cases. In this PR, at this point, we would know we need to search all occurrences of pre-existing `self.client` and then review each and every one to see whether/how to adjust it.

Reply:
I actually did the part where I looked through every occurrence of self.client and then decide if its is needed to be adjusted and I thought these changes are the minimun needed. However for cases you have suggested, if it does not really matter to initialize a client as we are pulling constant values, we can make such a change.

Reply:
So, does that mean the latest commit already address all the existing `self.client`? If you intended to leave some of them unchanged, we can discuss specifically. Right now I do not know whether those left-behind are intended or not.

Reply:
So, all the self.clients which were passed in functions as parameters which were called on self.client object were initially not changed as client object will get initialized before function parameters are evaluated. However, that can depend and vary on how python compiler processes a function call. To avoid this dependency, we realized after the offline conversation that all the places where self.client is still used are global variables and do not really depend on the specific instance of the class, so they have been changed to use Client.variable_name instead.